core/server/master/src/main/java/alluxio/master/DefaultSafeModeManager.java

```diff
@@ -30,47 +30,56 @@
   private final Clock mClock;
 
   /**
-   * Safe mode state. The value will be null if master is not in safe mode, or a time point in
-   * millisecond indicating when master will stop waiting for workers and leave safe mode.
+   * Safe mode state. The mark indicates whether master is in safe mode, the reference stores
+   * time point in millisecond indicating when master started waiting for workers.
    */
-  private AtomicReference<Long> mWorkerConnectWaitEndTimeMs = new AtomicReference<>();
+  private final AtomicMarkableReference<Long> mWorkerConnectWaitStartTimeMs =
+      new AtomicMarkableReference<>(null, true);
 
   /**
    * Creates {@link DefaultSafeModeManager} with default clock.
    */
   public DefaultSafeModeManager() {
-    this(new ElapsedTimeClock());
+    this(new ElapsedTimeClock(), true);
   }
 
   /**
    * Creates {@link DefaultSafeModeManager} with given clock.
    *
    * @param clock a {@link Clock} for calculating elapsed time
+   * @param enterSafeMode whether to start in safe mode
    */
-  public DefaultSafeModeManager(Clock clock) {
+  public DefaultSafeModeManager(Clock clock, boolean enterSafeMode) {
     mClock = clock;
+    mWorkerConnectWaitStartTimeMs.set(null, enterSafeMode);
   }
 
   @Override
   public void notifyRpcServerStarted() {
-    // updates end time after which master will leave safe mode
+    // updates start time when Alluxio master waits for workers to register
     long waitTime = Configuration.getMs(PropertyKey.MASTER_WORKER_CONNECT_WAIT_TIME);
     LOG.info(String.format("Entering safe mode. Expect leaving safe mode after %dms", waitTime));
-    mWorkerConnectWaitEndTimeMs.set(mClock.millis() + waitTime);
+    mWorkerConnectWaitStartTimeMs.set(mClock.millis(), true);
   }
 
   @Override
   public boolean isInSafeMode() {
-    // lazily updates safe mode state upon inquiry
-    Long endTime = mWorkerConnectWaitEndTimeMs.get();
-
     // bails out early before expensive clock checks
-    if (endTime == null) {
+    if (!mWorkerConnectWaitStartTimeMs.isMarked()) {
       return false;
     }
-    if (mClock.millis() < endTime) {
+
+    Long startTime = mWorkerConnectWaitStartTimeMs.getReference();
+    if (startTime == null) {
+      // master has not started waiting for workers yet
+      return true;
+    }
+
+    // lazily updates safe mode state upon inquiry
+    long waitTime = Configuration.getMs(PropertyKey.MASTER_WORKER_CONNECT_WAIT_TIME);
+    if (mClock.millis() - startTime < waitTime) {
       return true;
     }
-    return !mWorkerConnectWaitEndTimeMs.compareAndSet(endTime, null);
+    return !mWorkerConnectWaitStartTimeMs.compareAndSet(startTime, null, true, false);
```

Comment:
is this return value correct? If another thread comes in and brings us out of safe mode first, this will say we're still in safe mode.

Reply:
Good catch. This only works when one thread enter safe mode while another exit safe mode.  I will update it to return the actual value of the variable.